cypress-io/cypress
# feat: Selective CSP header stripping from HTTPResponse

- Closes [#1030](https://github.com/cypress-io/cypress/issues/1030)

### User facing changelog

- Cypress now allows selective stripping of `Content-Security-Policy` and `Content-Security-Policy-Report-Only` header directives via the `stripCspDirectives` config option.

### Additional details
- add generated `nonce` to inline script injection
- append generated `nonce` policy value to each CSP header `script-src-elem`, `script-src`, and `default-src` directive if provided in original response
- remove mandatory `content-security-policy` and `content-security-policy-report-only` header stripping
- selectively strip problematic CSP directive `frame-ancestors` because it prevents Cypress from loading target into iframe
- allows for proviiding CSP directives to strip in the `parseCspHeaders` method
- add config option `stripCspDirectives` that permits selective stripping of individual CSP directives
- default value for the `stripCspDirectives` config option maintains existing CSP header stripping

### Steps to test

1. Create a local HTTP server that responds to a request with CSP headers defined
2. Run a Cypress test that sends a request to the local server, and tests the response header, similar to the following:
```javascript
const route = 'http://localhost:8080';
cy.request(route).as('route');
cy.get('@route').should(response => {
    expect(response).to.have.property('headers');
    expect(response.headers).to.have.property('content-security-policy');
    expect(response.headers['content-security-policy']).to.match(
        /script-src 'nonce-[^-A-Za-z0-9+/=]|=[^=]|={3,}'/
    );
});
cy.visit(route);
```

### How has the user experience changed?
This change does not affect UI/UX

### PR Tasks
[Documentation for `stripCspDirectives`](https://github.com/cypress-io/cypress-documentation/pull/5194)
